### ISSUE TITLE
Settle on a Go dependency management solution

### ISSUE THREAD

**Issue by Shadowfiend** [keep-network/keep-core#33]
This started off in a playground repo... @keep-network/go please sound off/discuss :)

Currently two dimensions to this:
 - Which tool do we use? @l3x has a strong preference for Glide.
 - Do we commit vendored content? Alternatives include just committing lock files (or equivalents), and depending on keep-network clones of repos (so we are in complete control of the repos).

@Shadowfiend @ https://github.com/keep-network/go-experiments/pull/1#issuecomment-356955855:

> I would say our user is not the person who builds the Go code, it's the person who runs the Go code. If someone has issues building the Go code, that is directly our problem, and we should seek to fix it.
> 
> I'm interested in this argument:
> > it guards against upstream renames, deletes and commit history overwrites
> 
> I'm not sure I want to be guarded against that, insofar as I want to know when an upstream of ours is misbehaving in this fashion if we're depending on them. This immediately makes them less reliable, and should make us reconsider using them.
> 
> Tl;dr: the cons feel worse than the pros for me… It feels like a package manager that requires you to commit your vendor directory has simply failed as a package manager. But if everyone else feels like this is the way to go, let's do it. I will swallow my annoyance and live with 1.7 million LOC PRs that are secretly 78LOC PRs ;)
> 
> @keep-network/go Interested in thoughts from more of ya!
> 
> For the purposes of this PR, this discussion is moot. This is an experiments repo anyway, so it's the wild west and can just serve to inform the broader discussion. Merging!

@rargulati @ https://github.com/keep-network/go-experiments/pull/1#issuecomment-357039589:

> On the vendor thing - I agree with your annoyance. Go is a total disaster when it comes to package management - in the entirety of the thing. That being said, I’ve found that going with the conventions when starting out (vs going against them) saves a lot of headache. Buuuutttt it’s important to talk about these tradeoffs, because some conventions early on bite you really bad later (ie. ignoring package layout, monorepo vs many repo, etc)
> 
> The vendoring thing may bite us later - I don’t know, enough to confidently say it will. I've seen one Go vendoring solution at scale, and it ended up being bad - so dep is most definitely an improvement over that (bmizerany/vendor). I’ve got experience with Ruby (gems and bundling were ok) and Go (annoying, but not the worst). Friends point to cargo in Rust as a great solution.
> 
> For our client applications, it could make sense to revisit this.

@mhluongo @ https://github.com/keep-network/go-experiments/pull/1#issuecomment-357116927:

> From an in-person convo with @rargulati:
> 
> I wonder if we can get the best of both worlds by forking / hosting all dependencies, and only allowing vendoring from those forks? It means needing to touch another repo to update a dependency, but it also means we can't get leftpad'd or have someone else break the build, and it keeps PR's looking good.
> 
> Reproducible builds are important for this project, and that approach would kick it up a notch and make us focus on upstream changes.

Going to let Lex summarize his position from [this post](https://github.com/keep-network/go-experiments/pull/1#issuecomment-357134653), which provides a fairly detailed exploration of the problem for reference.

@lex @ https://github.com/keep-network/go-experiments/pull/1#issuecomment-364503887:

> After a bit more effort managing keep-core dependencies I created this Go dependency management document. It has screen shots and details. Below is the summary:
> 
>  - gx leverages ipfs, which is cool but it's in alpha and not ready for production use.
>  - dep is slightly better, but not by much. Feels like a beta and is time consuming.
>  - glide just works.
